bionic/flow.py

```diff
@@ -820,69 +896,75 @@ def __call__(self, func):
+
+        docs = acc.docs
+        orig_docstring = provider.get_source_func().__doc__
+        if docs is None:
+            if orig_docstring is not None and len(names) > 1:
+                preamble = """
+                Using a single doc string for a multi-output function is
+                deprecated and will become an error condition in a future
+                release; use the ``@docs`` decorator to specify
+                one doc string for each entity.  Details:
+                """
+                details = f"""
                 got {len(names)} names {tuple(names)!r} and
-                {len(docs)} docs {tuple(docs)!r}"""
-            )
-            if len(provider.attrs.docs) == 1:
-                prefix = oneline(
-                    """
-                    Using a single doc string for a multi-output function is
-                    deprecated and will become an error condition in a future
-                    release; use the ``@docs`` decorator to specify
-                    one doc string for each entity.  Details:"""
-                )
-                warnings.warn(prefix + "\n" + message)
+                just one docstring {orig_docstring}
+                """
+                warnings.warn(oneline(preamble) + "\n" + oneline(details))
 
-                docs = [docs[0]] * len(names)
-                provider = AttrUpdateProvider(
-                    provider, "docs", docs, allow_override=True
-                )
-            else:
-                raise ValueError(message)
+            docs = [orig_docstring for name in names]
+        if len(docs) != len(names):
+            message = f"""
+            Number of docs must match the number of names;
+            got {len(names)} names {tuple(names)!r} and
+            {len(docs)} docs {tuple(docs)!r}"""
+            raise ValueError(oneline(message))
+
+        can_persist = acc.can_persist
+        if can_persist is None:
+            can_persist = True
+
+        can_memoize = acc.can_memoize
+        if can_memoize is None:
+            can_memoize = True
+
+        if not can_persist and not can_memoize:
+            message = f"""
+            Attempted to set both @persist and @memoize to False.
+            At least one form of storage must be enabled for entities:
+            {provider.attrs.names!r}"""
+            raise ValueError(oneline(message))
+
+        entity_defs_by_name = {
+            name: EntityDefinition(
+                name=name,
+                protocol=protocol,
+                doc=doc,
+                can_persist=can_persist,
+                can_memoize=can_memoize,
+            )
+            for name, doc in zip(names, docs)
+        }
 
         state = self._state
 
+        # Delete the original definitions.
         state = state.delete_providers(provider.attrs.names)
+        state = state.delete_entity_defs(provider.attrs.names)
+
+        # Create the new definitions.
+        for name in provider.attrs.names:
```

Comment:
nit: I think this would be easier to read if you iterate over `entity_defs_by_name.values()` and define them rather than using `provider.attrs.names` for iteration.

Reply:
You're right!